samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -110,17 +108,73 @@ public ZkKeyBuilder getKeyBuilder() {
    * @return String representing the absolute ephemeralPath of this client in the current session
    */
   public synchronized String registerProcessorAndGetId(final ProcessorData data) {
+    String processorId = data.getProcessorId();
     if (ephemeralPath == null) {
-      ephemeralPath =
-          zkClient.createEphemeralSequential(
-              keyBuilder.getProcessorsPath() + "/", data.toString());
-
-      LOG.info("newly generated path for " + data +  " is " +  ephemeralPath);
-      return ephemeralPath;
+      /**
+       * If there is an existing processor with same processorId as incoming processorId, throw up and fail.
+       * Caller of this api should handle exception and kill the stream processor.
+       */
+      ephemeralPath = zkClient.createEphemeralSequential(keyBuilder.getProcessorsPath() + "/", data.toString());
+      LOG.info("Created ephemeral path: {} for processor: {} in zookeeper.", ephemeralPath, data);
+      ProcessorNode processorNode = new ProcessorNode(data, ephemeralPath);
+      // Determine if there are duplicate processors with this.processorId after registration.
+      if (shouldStopProcessor(processorNode)) {
+        LOG.info("Processor: {} is duplicate. Deleting zookeeper node at path: {}.", processorId, ephemeralPath);
+        zkClient.delete(ephemeralPath);
+        throw new SamzaException(String.format("Processor: %s is duplicate in the group. Registration failed.", processorId));
+      }
     } else {
-      LOG.info("existing path for " + data +  " is " +  ephemeralPath);
-      return ephemeralPath;
+      LOG.info("Ephemeral path: {} exists for processor: {} in zookeeper.", ephemeralPath, data);
     }
+    return ephemeralPath;
+  }
+
+  /**
+   * Determines if a processor should be stopped.
```

Comment:
@navina

When there’re multiple duplicate processors(P1,P2,P3..Pn) this marks processor with lowest sequential id as not duplicate, rest as duplicate.

I don’t think this  is synonymous to `Collection.contains` to rename(which should be symmetric for all processors). If we rename that way, it also conveys wrong meaning that we kill all duplicate processors (P1, P2, .. Pn). 

Naming to  `isValidRegisteredProcessor` if it's okay